src/models/ontology_models.py

```diff
@@ -0,0 +1,118 @@
+from transformers import AutoTokenizer, AutoModel, AutoModelForMaskedLM
+from tqdm.auto import tqdm
+import torch
+from abc import ABC, abstractmethod
+import yaml
+
+class OntoModelsBase:
+    def __init__(self, method:str, query:list, corpus:list, yaml_path:str='method_model.yaml') -> None:    
+        self.method = method
+        self.query = query
+        self.corpus = corpus
+        if self.method is None:
+            raise ValueError("Method name cannot be None")
+        
+        if len(self.query) == 0:
+            raise ValueError("Query list cannot be empty")
+        if len(self.corpus) == 0:
+            raise ValueError("Corpus list cannot be empty")
+        
+        # Load method_model_dict from a YAML file
+        self.method_model_dict = self.load_method_model_dict(yaml_path)
+        self.list_of_methods = list(self.method_model_dict.keys())
+        if self.method not in self.list_of_methods:
+            raise ValueError(f"Method name should be one of {self.list_of_methods}")
+        self.matches_tmp = {
+            "original_value": [],
+            "curated_ontology": [],
+            "match_level": [],
+            "top1_match": [],
+            "top1_score": [],
+            "top2_match": [],
+            "top2_score": [],
+            "top3_match": [],
+            "top3_score": [],
+            "top4_match": [],
+            "top4_score": [],
+            "top5_match": [],
+            "top5_score": [],
+        }
+        
+
+        self.topk = 5 
+    
+    def load_method_model_dict(self, file_path):
+        """
+        This method is responsible for loading the method_model_dict from a YAML file.
+        
+        Args:
+            file_path (str): The path to the YAML file.
+        
+        Returns:
+            dict: The dictionary containing the method names and their corresponding model names.
+        """
+        with open(file_path, 'r') as file:
+            # Load the dictionary from YAML file
+            method_model_dict = yaml.safe_load(file)
+        return method_model_dict
+
+    @abstractmethod
+    #Mean Pooling - Take attention mask into account for correct averaging
+    def mean_pooling(self, model_output, attention_mask):
+        """
+        Performs mean pooling on the token embeddings.
+
+        Args:
+            model_output (tuple): The output of the model, where the first element contains all token embeddings.
+            attention_mask (torch.Tensor): The attention mask for the input tokens.
+
+        Returns:
+            torch.Tensor: The mean-pooled token embeddings.
+
+        """
+        token_embeddings = model_output[0] #First element of model_output contains all token embeddings
+        input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
+        return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)
+
+    @abstractmethod
```

Comment:
I think you can take these decorators out too - this class is a base class, but in general not an abstract base class as for how its used

Reply:
I see - i'll just remove the @abstractmethod from these. Thanks!